src/plugins/kibana_utils/common/persistable_state/index.ts

```diff
@@ -47,8 +52,29 @@ export interface PersistableState<P extends SerializableState = SerializableStat
    * @param state
    */
   extract: (state: P) => { state: P; references: SavedObjectReference[] };
+
+  /**
+   * migrateToLatest function receives state of older version and should migrate to the latest version
+   * @param state
+   * @param version
+   */
+  migrateToLatest?: (state: SerializableState, version: string) => P;
+
+  /**
+   * migrate function runs the specified migration
+   * @param state
+   * @param version
+   */
+  migrate?: (state: SerializableState, version: string) => SerializableState;
```

Comment:
I'm wondering if it would help clean up some of the casting if we provided a `FromVersion` and `ToVersion` here as well (as in `MigrateFunction`). Then in executor you wouldn't need to cast to `ExpressionAstFunction` and could instead provide it in the generic type params.

Reply:
i think we should keep this one as it is, as we want to keep types for old versions of state internal (nobody should be messing with that) and migrate really doesn't know anything more about the state coming in apart that it needs to be serializable. (nor does the consumer)